### ISSUE TITLE
gff to gtf conversion

### ISSUE THREAD

**Issue by rbadmi** [rbadmi/Troubleshooting#1]
Here I will be describing the problems I encountered while using the above link and how I solved them. These may seem pretty obvious for people who are aware, but I am sure a new person who is still learning would find these incredibly helpful.

### Option 1: Using RSEM: https://github.com/deweylab/RSEM
You should install RSEM first as in the above link. More about it in my RSEM article.

https://manpages.ubuntu.com/manpages/jammy/man1/rsem-gff3-to-gtf.1.html

The above link shows one of the ways to convert a gff3 annotation file to gtf file using RSEM (actually it is just one of the tool in the entire RSEM package). This is required as several tools need a gtf file for building references, mapping the reads to a genome or counting the reads for your RNA-seq data.

1. the code as in the link: rsem-gff3-to-gtf input_GFF3_file Input GFF3 file. output_GTF_file Output GTF file.
2. the code as it makes sense for a newbie: `rsem-gff3-to-gtf your_gff_filename your_gtf_filename_thatyouwant_afterconversion`

**Problem 1:**
1. If you open your gff3 file (in Notepad++), you will see different attributes like gene, mRNA, 5'-UTR, 3'-UTR and exon. But once you convert to gtf file you will lose all these and the gtf file will only retain the exon attributes with this method
2. I have tried AGAT which according to this link (https://agat.readthedocs.io/en/latest/gff_to_gtf.html) preserves all the attributes in gtf file. But after a lot of troubleshooting, I was unable to get it to work as the remote server I am using was not supporting the dependencies AGAT would need. I am still trying to solve this when I get some time.

**Problem 2:**
1. In my gff3 file, there were empty lines incorporated after every 6 lines or every 12 lines of annotations. You will have to open your gff3 file (Notepad++) and see. So, when I ran the command (rsem-gff3-to-gtf your_gff_filename your_gtf_filename_thatyouwant_afterconversion), I was getting an error saying Line 7 does not have 9 fields. I presume the 9 fields it is referring to are the 9 columns in my gff3 file.
2. This is what I did to remove those blank lines: In Notepad++ Select Edit> Line Operations > Remove Empty Lines( Containing Blank Characters) and removed all the empty lines.
3. I saved this file and used it to run the above command and it worked.

### Option 2: Using Cufflinks: http://cole-trapnell-lab.github.io/cufflinks/getting_started/#install-quick-start
You should install Cufflinks from the above link

1. Cufflinks can be installed by just unzipping the tar.gz file as mentioned in the above link `tar -zxvf cufflinks-2.2.1.Linux_x86_64.tar.gz`
2. Once unzipped cd into the folder (go into the folder by using cd command) and make sure you have cufflinks, cuffdiff and cuffcompare folders in the main cufflinks folder. Then, export the main folder’s path as below: `export PATH=$PATH:/home/raghu/cufflinks-2.2.1.Linux_x86_64`. Change this path according to your own folder names
3. Once you have successfully done this, the installation is complete
4. You can test the installation by downloading the test data and running `cufflinks ./test_data.sam` as described in the above link
5. The actual code is `gffread -E -O -T input.gff -o output.gtf`, but this does not preserve all the attributes
6. So, in this link (https://www.biostars.org/p/331485/), someone suggested to use `gffread -E -F -O -T input.gff -o output.gtf` to preserve the attributes. This worked for me, but I can only see exon and CDS attributes in my gtf file, whereas in gff3 file I have gene, mRNA and UTR regions as well.
7. Description of the options: -E option directs gffread to "expose" (display warnings about) any potential issues encountered while parsing the input file, -F option is to preserve all original GFF attributes (for non-exon features) but repetitive/redundant
exon/CDS attributes are merged into the parent transcript attributes and -T option means that the output format will be gtf2. You can read more about it in the article (https://f1000research.com/articles/9-304/v1#T2) or this link: http://ccb.jhu.edu/software/stringtie/gff.shtml#gffread
8. One more thing I noticed is that gffread in Cufflinks worked perfectly well even when I used the gff3 file with blank spaces described above in Option 1.
